SecuritySystemUWP/SecuritySystemUWP/Storage/OneDrive.cs

```diff
@@ -158,5 +154,98 @@ private void SaveTokens(object sender, string arg)
             App.Controller.XmlSettings.OneDriveAccessToken = oneDriveConnector.accessToken;
             App.Controller.XmlSettings.OneDriveRefreshToken = oneDriveConnector.refreshToken;
         }
+
+        private async Task uploadWithRetry(StorageFile file, int tryNumber)
+        {
+            HttpResponseMessage response = await oneDriveConnector.UploadFileAsync(file, String.Format("{0}/{1}", App.Controller.XmlSettings.OneDriveFolderPath, DateTime.Now.ToString("yyyy_MM_dd")));
+            bool success = await parseResponse(response, tryNumber);
+            var events = new Dictionary<string, string>();
+
+            if (success)
+            {
+                numberUploaded++;
+                await file.DeleteAsync();
+                this.lastUploadTime = DateTime.Now;
+            }
+            else if (tryNumber <= MaxTries)
+            {
+                events.Add("Retrying upload", tryNumber.ToString());
+                TelemetryHelper.TrackEvent("FailedToUploadPicture", events);
+                await uploadWithRetry(file, ++tryNumber);
+            }
+            else
+            {
+                events.Add("Max upload attempts reached", tryNumber.ToString());
+                TelemetryHelper.TrackEvent("FailedToUploadPicture - total failure", events);

```

Comment:
Another thought: how hard is it to sort the Events based on the event properties added? This way, ,the main telemetry event for all of this could stay as "FailedToUploadPicture", and the property in each could have the more descriptive text.

Ultimately, I'll be for whatever way it makes it easiest for you to parse the telemetry on the other side :)